proposals/0000-extending-capi-calling-convention.rst

```diff
@@ -0,0 +1,96 @@
+Both ccall and capi disallow passing aggregated types directly, but for capi this is not strictly required as it already produces wrapper functions in C. 
+
+Proposed Change Specification
+-----------------------------
+The proposed changes affect how capi calls are handled. 
+The parameters and output of the called C function can be classified into three categories: 
+1. primitives and newtypes around primitives (example: CInt, CFloat)
+2. newtypes around primitives declared with a CTYPE pragma 
+3. data types implementing a Storable instance declared with a CTYPE pragma
+
+A parameter of category 1 is passed and returned as is already implemented. 
+
+For category 2, the parameter is passed by the value of the encase primitive like category 1. This category corrisponds with C structs that possess only one member of a primitive type (example: 
+``typedef struct { int value } unique_id;``). In the C wrapper function, the corrisponding parameter has the member type and is encapsulated in a struct before being passed to the wrapped function.
+For return value in this category, the value is returned in the primitive type of the singleton member. 
+
+Parameters of category 3 are passed by pointers. The wrapper function shalls dereference these pointers before passing them to the wrapped function.  
+If the return value is in this category, the wrapper function shalls have one extra rightmost pointer parameter where the result of the wrapped function will be written to. 
+The types of category 3 must implement a Storable instance in order for Haskell to marshal them to and from pointers. 
+
+Examples
+--------
+Suppose the the C function ``aggregate wrapped(int, aggregate)`` is needed to be called from Haskell. This would be how it is done. 
+
+    data {-# CTYPE "aggregate" #-} Aggregate = ...
+    instance Storable Aggregate where ...
+ 
+    foreign unsafe capi "header.h wrapped" ffunc :: CInt -> Aggregate -> Aggregate
+
+Generated C wrapper function
+
+    void wrapper (int _0, aggregate \*_1, aggregate \*_result) { \*_result = wrapped(_0, \*_1); }
+
+Equivalent in ccall convention
+
+    foreign unsafe ccall "wrapper" ffunc_wrapper :: CInt -> Ptr Aggregate -> Ptr Aggregate -> IO ()
+    ffunc :: CInt -> Aggregate -> Aggregate
+    ffunc _0 _1 = unsafePerformIO (with _1 (\ _1' -> alloca (\ _result -> ffunc_wrapper _0 _1 _result >> peek _result)))
+
+
+For cases of category 2
+
+    newtype {-# CTYPE "unique_id" #-} UniqueId = UniqueId CInt
+    
+    foreign unsafe capi "header.h wrapped" ffunc :: UniqueId -> Ptr () -> IO UniqueId 
+
+C wrapper function
+
+    int wrapper (int _0, void \*_1) { unique_id result = wrapped((unique_id){_0}, _1); return *(int*)&result; }
+
+ccall equivalent
+
+    foreign unsafe ccall "wrapper" ffunc_wrapper :: UniqueId -> Ptr () -> IO UniqueId
+    ffunc :: UniqueId -> Ptr () -> IO UniqueId
+    ffunc _0 _1 = ffunc_wrapper _0 _1
```

Comment:
```suggestion
Suppose the the C function ``aggregate wrapped(int, aggregate)`` is needed to be called from Haskell. This would be how it is done::

    data {-# CTYPE "aggregate" #-} Aggregate = ...
    instance Storable Aggregate where ...
 
    foreign unsafe capi "header.h wrapped" ffunc :: CInt -> Aggregate -> Aggregate

Generated C wrapper function::

    void wrapper (int _0, aggregate \*_1, aggregate \*_result) { \*_result = wrapped(_0, \*_1); }

Equivalent in ccall convention::

    foreign unsafe ccall "wrapper" ffunc_wrapper :: CInt -> Ptr Aggregate -> Ptr Aggregate -> IO ()
    ffunc :: CInt -> Aggregate -> Aggregate
    ffunc _0 _1 = unsafePerformIO (with _1 (\ _1' -> alloca (\ _result -> ffunc_wrapper _0 _1 _result >> peek _result)))


For cases of category 2::

    newtype {-# CTYPE "unique_id" #-} UniqueId = UniqueId CInt
    
    foreign unsafe capi "header.h wrapped" ffunc :: UniqueId -> Ptr () -> IO UniqueId 

C wrapper function::

    int wrapper (int _0, void \*_1) { unique_id result = wrapped((unique_id){_0}, _1); return *(int*)&result; }

ccall equivalent::

    foreign unsafe ccall "wrapper" ffunc_wrapper :: UniqueId -> Ptr () -> IO UniqueId
    ffunc :: UniqueId -> Ptr () -> IO UniqueId
    ffunc _0 _1 = ffunc_wrapper _0 _1
```
This way, the code will be presented in code blocks.